Migrate PKDateComponentsRange from WKKeyedCoder

sheeparegreat opened the pull request:
#### 9a9cd757efb1bebac4626bdffed9124d75b43671
<pre>
Migrate PKDateComponentsRange from WKKeyedCoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=300141">https://bugs.webkit.org/show_bug.cgi?id=300141</a>
<a href="https://rdar.apple.com/137149750">rdar://137149750</a>

Reviewed by Abrar Rahman Protyasha.

PKDateComponentsRange did not need to use WKKeyedCoder it had enough API for serialization.

Covered by existing tests.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.h: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.mm: Added.
(WebKit::CoreIPCPKDateComponentsRange::CoreIPCPKDateComponentsRange):
(WebKit::CoreIPCPKDateComponentsRange::toID const):
(IPC::encodeObjectDirectly&lt;PKDateComponentsRange&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;PKDateComponentsRange&gt;):
* Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPassKit.serialization.in:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/301207@main">https://commits.webkit.org/301207@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb66e2d1ed0196bbb1bf9d287cebe88e122ebc54

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125063 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44731 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35469 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131911 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77055 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/574a8031-645b-4c26-88e3-52547256c2d2) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45422 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53296 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95200 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1f2f809a-aabd-4e2c-82f4-32baa4aca3db) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128017 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36256 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111849 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75742 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8cfdabdd-ea7e-47ca-adef-2e65fc266684) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30003 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75393 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117158 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106020 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30232 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134589 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123581 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51881 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39676 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103670 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52308 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108061 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103442 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26381 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48787 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27078 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48931 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51768 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57560 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156604 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51140 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39224 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54497 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52831 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

adetaylor commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -2469,6 +2469,8 @@
 		F454C12C2BA3AC8C00871551 /* _WKTargetedElementInfoInternal.h in Headers */ = {isa = PBXBuildFile; fileRef = F454C12B2BA3AC8C00871551 /* _WKTargetedElementInfoInternal.h */; };
 		F454C1302BA3AFFD00871551 /* APITargetedElementInfo.h in Headers */ = {isa = PBXBuildFile; fileRef = F454C12D2BA3AFFD00871551 /* APITargetedElementInfo.h */; };
+		F458B8EA2E8F46E900948328 /* CoreIPCPKDateComponentsRange.h in Headers */ = {isa = PBXBuildFile; fileRef = F458B8E72E8F46E100948328 /* CoreIPCPKDateComponentsRange.h */; };
+		F458B8EB2E8F46F000948328 /* CoreIPCPKDateComponentsRange.mm in Sources */ = {isa = PBXBuildFile; fileRef = F458B8E82E8F46E100948328 /* CoreIPCPKDateComponentsRange.mm */; };
```
Should you be adding this as input to the unified build concatenation process instead? I'm not sure of the guidelines, or whether this necessarily results in a standalone build step here

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.h:
```diff
@@ -0,0 +1,60 @@
+
+#include "CoreIPCDateComponents.h"
+
+#include <wtf/ArgumentCoder.h>
```
Remove?

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.mm:
```diff
@@ -0,0 +1,73 @@
+#if USE(PASSKIT) && HAVE(WK_SECURE_CODING_PKDATECOMPONENTSRANGE)
+
+#import "ArgumentCoders.h"
+#import "ArgumentCodersCocoa.h"
```
Remove?

aprotyas commented on Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.mm:
```diff
@@ -0,0 +1,73 @@
+    auto result = decoder.decode<std::optional<WebKit::CoreIPCPKDateComponentsRange>>();
+    if (!result)
+        return std::nullopt;
+    return *result ? (*result)->toID() : nullptr;
```
This is a good use case for [std::optional::and_then](https://en.cppreference.com/w/cpp/utility/optional/and_then.html)

```suggestion
    return result.and_then([](RetainPtr<id> object) {
        return object ? object->toID() : nullptr;
    });
```

aprotyas commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
```diff
@@ -484,6 +484,11 @@ NSURLRequest wrapped by CoreIPCNSURLRequest
 AVOutputContext wrapped by CoreIPCAVOutputContext
 #endif
+
+#if USE(PASSKIT) && !HAVE(WK_SECURE_CODING_PKDATECOMPONENTSRANGE)
+PKDateComponentsRange wrapped by CoreIPCPKDateComponentsRange
+#endif
```
I think I'm misunderstanding how this works, but wouldn't we want `wrapped by CoreIPCPKDateComponentsRange` if `HAVE(WK_SECURE_CODING_PKDATECOMPONENTSRANGE)`?

And if `!HAVE(WK_SECURE_CODING_PKDATECOMPONENTSRANGE)`, do we not want to `ENCODE_AS_SECURE_CODING(PKDateComponentsRange)`?

sheeparegreat replied:
In the current shipping case we have `PKDateComponentsRange wrapped by CoreIPCPKDateComponentsRange` and 

`[WebKitSecureCodingClass=PAL::getPKDateComponentsRangeClassSingleton(), SupportWKKeyedCoder, WebKitPlatform] webkit_secure_coding PKDateComponentsRange {
    startDateComponents: NSDateComponents
    endDateComponents: NSDateComponents
}`

With this change we switch to use the explicit code I added in this change

aprotyas commented on Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.mm:
```diff
@@ -0,0 +1,75 @@
+    encoder << (instance ? std::optional(WebKit::CoreIPCPKDateComponentsRange(instance)) : std::nullopt);
+}
+
+template<> std::optional<RetainPtr<id>> decodeObjectDirectlyRequiringAllowedClasses<PKDateComponentsRange>(Decoder& decoder)
+{
+    return decoder.decode<std::optional<WebKit::CoreIPCPKDateComponentsRange>>()
+        .and_then([](const std::optional<WebKit::CoreIPCPKDateComponentsRange>& inner) -> std::optional<RetainPtr<id>> {
+            return inner ? inner->toID() : nullptr;
+        });
+}
```
If you wanted to continue with the monadic unwrapping 😄 

```suggestion
template<> std::optional<RetainPtr<id>> decodeObjectDirectlyRequiringAllowedClasses<PKDateComponentsRange>(Decoder& decoder)
{
    return decoder.decode<std::optional<WebKit::CoreIPCPKDateComponentsRange>>()
        .and_then([](const std::optional<WebKit::CoreIPCPKDateComponentsRange>& inner) -> std::optional<RetainPtr<id>> {
            return inner.and_then([](const WebKit::CoreIPCPKDateComponentsRange& range){
                return range.toID();
            }).or_else({
                return std::make_optional(nullptr);
            });
        });
}
```

sheeparegreat committed:
Migrate PKDateComponentsRange from WKKeyedCoder
https://bugs.webkit.org/show_bug.cgi?id=300141
rdar://137149750

Reviewed by Abrar Rahman Protyasha.

PKDateComponentsRange did not need to use WKKeyedCoder it had enough API for serialization.

Covered by existing tests.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.h: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.mm: Added.
(WebKit::CoreIPCPKDateComponentsRange::CoreIPCPKDateComponentsRange):
(WebKit::CoreIPCPKDateComponentsRange::toID const):
(IPC::encodeObjectDirectly<PKDateComponentsRange>):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<PKDateComponentsRange>):
* Source/WebKit/Shared/Cocoa/CoreIPCPKDateComponentsRange.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPassKit.serialization.in:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/301207@main
